sw/device/lib/dif/dif_i2c.h

```diff
@@ -0,0 +1,386 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_I2C_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_I2C_H_
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+/**
+ * Represents a speed setting for an I2C component: standard, fast, and
+ * fast plus, corresponding to 100 kbaud, 400 kbaud, and 1 Mbaud,
+ * respectively.
+ *
+ * The integer values of the variants are equal to the maximum speed of
+ * that mode, in kbaud.
+ */
+typedef enum dif_i2c_speed {
+  kDifI2cBusSpeedStandard = 100,
+  kDifI2cBusSpeedFast = 400,
+  kDifI2cBusSpeedFastPlus = 1000,
+} dif_i2c_speed_t;
+
+/*
+ * Timing parameters for an I2C device. Actually computing these values is
+ * somewhat complicated, so this struct should be assembled using the
+ * `dif_i2c_compute_timing()` function. A caller is, however, free to assemble
+ * this struct or modify a return value of `dif_i2c_compute_timing()`, so long
+ * as the spec is respected.
+ *
+ * These values correspond to those in Table 10 of the I2C spec, and are given
+ * in units of input clock cycles.
+ */
+typedef struct dif_i2c_timing_params {
+  uint16_t pulse_time_high;
+  uint16_t pulse_time_low;
+  uint16_t rise_time;
+  uint16_t fall_time;
+  uint16_t start_signal_setup_time;
+  uint16_t start_signal_hold_time;
+  uint16_t data_signal_setup_time;
+  uint16_t data_signal_hold_time;
+  uint16_t stop_signal_setup_time;
+  uint16_t stop_signal_hold_time;
+} dif_i2c_timing_params_t;
+
+/**
+ * Configuration for computing a value of `dif_i2c_timing_params`, which itself
+ * is used to configure an I2C device.
+ */
+typedef struct dif_i2c_timing_config {
+  /**
+   * Indicates the lowest speed at which an I2C target connected to this host
+   * will operate.
+   */
+  dif_i2c_speed_t min_speed;
+  /**
+   * Indicates the period of the clock driving this device, in nanoseconds.
+   *
+   * This value should not be zero, since it is used as a divisor for
+   * division.
+   */
+  uint32_t clock_period_nanos;
+  /**
+   * Indicates the expected time it takes for the I2C bus signal to rise, in
+   * nanoseconds. This value is dependent on properties of the hardware's
+   * interconnect, and not under actual firmware control.
+   */
+  uint32_t signal_rise_nanos;
+  /**
+   * Indicates the expected time for the bus singal to fall, similar to
+   * `signal_rise_nanos`.
+   */
+  uint32_t signal_fall_nanos;
+  /**
+   * Indicates the desired period of the SCL line, in nanoseconds. Normally,
+   * this should just be `1'000'000 / min_speed`, but the period may be
+   * larger if desired.
+   *
+   * Setting this value to zero will result in the minimum period being used.
+   */
+  uint32_t scl_period_nanos;
```

Comment:
Mixed terms of `signal` and `scl` here. I'd use `scl` and `sda`.

Reply:
Uh. Should signal just be `scl` then? I don't think any of these want to be `sda`.

Reply:
scl is SPI clock, sda is SPI data. So signal should be sda . I'd keep these in SPI terms, so scl / sda.

Reply:
Can you address the signal/sda comment please.